flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -245,22 +255,26 @@ public TaskExecutorGateway call() throws Exception {
 						resourceID, taskExecutorAddress, leaderSessionID, resourceManagerLeaderId);
 					throw new Exception("Invalid leader session id");
 				}
-				return getRpcService().connect(taskExecutorAddress, TaskExecutorGateway.class).get(5, TimeUnit.SECONDS);
+				return getRpcService().connect(taskExecutorAddress, TaskExecutorGateway.class)
+					.get(timeout.toMilliseconds(), timeout.getUnit());
 			}
 		}).handleAsync(new BiFunction<TaskExecutorGateway, Throwable, RegistrationResponse>() {
 			@Override
 			public RegistrationResponse apply(TaskExecutorGateway taskExecutorGateway, Throwable throwable) {
 				if (throwable != null) {
 					return new RegistrationResponse.Decline(throwable.getMessage());
 				} else {
-					WorkerType oldWorker = taskExecutorGateways.remove(resourceID);
-					if (oldWorker != null) {
+					WorkerRegistration oldRegistration = taskExecutors.remove(resourceID);
+					if (oldRegistration != null) {
 						// TODO :: suggest old taskExecutor to stop itself
 						slotManager.notifyTaskManagerFailure(resourceID);
```

Comment:
notifyTaskManagerFailure is called again in slotManager.registerTaskExecutor, maybe we should only keep one of these